crates/cubecl-wgpu/src/compute/server.rs

```diff
@@ -308,69 +249,79 @@ where
         let pipeline = self.pipeline(kernel, mode);
         let group_layout = pipeline.get_bind_group_layout(0);
 
-        let memory_handles = bindings
-            .into_iter()
-            .map(|binding| self.memory_management.get(binding.memory))
-            .collect::<Vec<_>>();
+        // Store all the resources we'll be using. This could be eliminated if
+        // there was a way to tie the lifetime of the resource to the memory handle.
+        let resources: Vec<_> = bindings
+            .iter()
+            .map(|binding| {
+                let resource_handle = self.memory_management.get(binding.memory.clone());
+                // Keep track of the storage we've used so far.
+                self.compute_storage_used.push(resource_handle.id);
+
+                self.memory_management.storage().get(&resource_handle)
+            })
+            .collect();
 
-        let entries = memory_handles
+        let entries = &resources
             .iter()
             .enumerate()
-            .map(|(i, buffer)| wgpu::BindGroupEntry {
+            .map(|(i, r)| wgpu::BindGroupEntry {
                 binding: i as u32,
-                resource: buffer.as_binding(),
+                resource: r.as_binding(),
             })
             .collect::<Vec<_>>();
 
         let bind_group = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
             label: None,
             layout: &group_layout,
-            entries: &entries,
+            entries,
         });
 
-        self.register_compute(pipeline, bind_group, count);
+        // First resolve the dispatch buffer if needed. The weird ordering is because the lifetime of this
+        // needs to be longer than the compute pass, so we can't do this just before dispatching.
+        let dispatch_resource = match count.clone() {
+            CubeCount::Dynamic(binding) => Some(self.get_resource(binding)),
+            _ => None,
+        };
+
+        let pass = self.start_compute_pass();
```

Comment:
I think I would rename `start_compute_pass` and `end_compute_pass` to `add_compute_pass` and `clear_compute_pass`, otherwise it's strange that we start way more compute pass than we end.